ipn/ipnauth/win.go

```diff
@@ -0,0 +1,205 @@
+	if res := checker.mustGrant(ServePath, wi.checkElevatedAdmin); res.HasResult() {
+		return res
+	}
+
+	// Profile owners have unrestricted access to their own profiles.
+	if wi.isProfileOwner(profile) {
+		if res := checker.grant(UnrestrictedProfileAccess); res.HasResult() {
+			return res
+		}
+	}
+
+	if wi.isProfileUser(profile, prefs) {
+		// Allow profile users to read basic profile info (e.g. profile and tailnet name) and list network devices.
+		profileUserRights := ReadProfileInfo | ListPeers
+		if !wi.env.IsServer && !wi.env.IsManaged {
+			// Additionally, on non-managed Windows client devices we should allow profile users to
+			// connect / disconnect, read preferences and select exit nodes.
+			profileUserRights |= Connect | Disconnect | ReadPrefs | ChangeExitNode
+		}
+		if res := checker.grant(profileUserRights); res.HasResult() {
+			return res
+		}
+	}
+
+	if !wi.env.IsServer && !isProfileEnforced(profile, prefs) {
+		// Allow any user to disconnect from non-enforced Tailnets on non-Windows Server devices.
+		// TODO(nickkhyl): automatically disconnect from the current Tailnet
+		// when a different user logs in or unlocks their Windows session,
+		// unless the unattended mode is enabled. But in the meantime, we should allow users
+		// to disconnect themselves.
+		if res := checker.grant(Disconnect); res.HasResult() {
+			return res
+		}
+	}
+
+	if isAdmin, _ := wi.tok.IsAdministrator(); isAdmin {
+		// Allow local admins to disconnect from any tailnet.
+		localAdminRights := Disconnect
+		if wi.env.IsServer || wi.env.IsManaged || wi.tok.IsElevated() {
+			// Allow elevated admins, as well as all admins on managed and server devices,
+			// unrestricted access to all local profiles, except for reading private keys.
+			localAdminRights |= UnrestrictedProfileAccess & ^ReadPrivateKeys
+		}
+		if res := checker.grant(localAdminRights); res.HasResult() {
+			return res
+		}
+	}
+
+	return checker.result()
+}
+
+// Close implements io.Closer by releasing resources associated with the Windows user identity.
+func (wi *WindowsIdentity) Close() error {
+	if wi == nil || wi.tok == nil {
+		return nil
+	}
+	if err := wi.tok.Close(); err != nil {
+		return err
+	}
+	runtime.SetFinalizer(wi, nil)
+	wi.tok = nil
+	return nil
+}
+
+func (wi *WindowsIdentity) checkAdmin() error {
+	isAdmin, err := wi.tok.IsAdministrator()
+	if err != nil {
+		return err
+	}
+	if !isAdmin {
+		return errors.New("the requested operation requires local admin rights")
+	}
+	return nil
+}
+
+func (wi *WindowsIdentity) checkElevatedAdmin() error {
+	if !wi.tok.IsElevated() {
+		return errors.New("the requested operation requires elevation")
+	}
+	return nil
+}
+
+func (wi *WindowsIdentity) isProfileOwner(profile ipn.LoginProfileView) bool {
+	return wi.tok.IsUID(profile.LocalUserID())
+}
+
+func (wi *WindowsIdentity) isProfileUser(profile ipn.LoginProfileView, prefs ipn.PrefsGetter) bool {
```

Comment:
document what a "profile user" is?

Reply:
Done. I also decided to make it a free function and rename it to `isProfileShared`.

It may or may not be a good idea to allow sharing profiles with specific users, rather than with all users on a device, and without enabling unattended mode. However, currently, this is a characteristic of the profile rather than of the user.